File: rust/crates/baa-native/src/image.rs
//! Reading a `.fleece` image.
//!
//! The writer lives in `src/native/image.ts`. The two are one format described
//! twice, which is a real risk, so three things keep them honest: the magic
//! and version are checked before anything is decoded, every read is bounds
//! checked and returns an error rather than panicking, and the test suite
//! builds an image from the TypeScript side and runs it through this reader on
//! every commit.
//!
//! A malformed image is refused, not guessed at. This code runs on whatever
//! bytes are appended to the executable, so it treats them as untrusted.

use std::rc::Rc;

use crate::ast::{
    Binding, BinaryOp, Block, Else, Expr, FnDecl, Image, ImportSpec, ImportTarget, LogicalOp,
    MatchArm, Module, Param, Pattern, Span, Stmt, StringPart, UnaryOp,
};

pub const MAGIC: &[u8] = b"FLEECE\n";
pub const VERSION: u8 = 1;

pub struct Reader<'a> {
    bytes: &'a [u8],
    at: usize,
    strings: Vec<Rc<str>>,
}

pub type Result<T> = std::result::Result<T, String>;

impl<'a> Reader<'a> {
    fn u8(&mut self) -> Result<u8> {
        let byte = *self.bytes.get(self.at).ok_or("image ends in the middle of a value")?;
        self.at += 1;
        Ok(byte)
    }

    fn u32(&mut self) -> Result<u32> {
        let end = self.at + 4;
        let slice = self.bytes.get(self.at..end).ok_or("image ends in the middle of a number")?;
        self.at = end;
        Ok(u32::from_le_bytes([slice[0], slice[1], slice[2], slice[3]]))
    }

    fn f64(&mut self) -> Result<f64> {
        let end = self.at + 8;
        let slice = self.bytes.get(self.at..end).ok_or("image ends in the middle of a number")?;
        self.at = end;
        let mut buffer = [0u8; 8];
        buffer.copy_from_slice(slice);
        Ok(f64::from_le_bytes(buffer))
    }

    fn bool(&mut self) -> Result<bool> {
        Ok(self.u8()? != 0)
    }

    fn str(&mut self) -> Result<Rc<str>> {
        let index = self.u32()? as usize;
        self.strings
            .get(index)
            .cloned()
            .ok_or_else(|| format!("string {index} is not in the image"))
    }

    fn span(&mut self) -> Result<Span> {
        Ok(Span { start: self.u32()?, end: self.u32()? })
    }

    fn count(&mut self) -> Result<usize> {
        let count = self.u32()? as usize;
        // A length is four bytes and an element is at least one, so a count
        // larger than what remains is corruption rather than a big program.
        // Without this a bad length asks for a several-gigabyte allocation.
        if count > self.bytes.len() - self.at.min(self.bytes.len()) + 1 {
            return Err(format!("image claims {count} items, which cannot fit in it"));
        }
        Ok(count)
    }

    fn list<T>(&mut self, mut read: impl FnMut(&mut Self) -> Result<T>) -> Result<Vec<T>> {
        let count = self.count()?;
        let mut out = Vec::with_capacity(count);
        for _ in 0..count {
            out.push(read(self)?);
        }
        Ok(out)
    }

    fn params(&mut self) -> Result<Rc<[Param]>> {
        let list = self.list(|r| {
            let name = r.str()?;
            let rest = r.bool()?;
            let default = if r.u8()? == 0 { None } else { Some(r.expr()?) };
            Ok(Param { name, rest, default })
        })?;
        Ok(Rc::from(list))
    }

    fn block(&mut self) -> Result<Rc<Block>> {
        Ok(Rc::new(self.list(|r| r.stmt())?))
    }

    fn binding(&mut self) -> Result<Binding> {
        let span = self.span()?;
        match self.u8()? {
            1 => Ok(Binding::Name { span, name: self.str()? }),
            2 => {
                let elements = self.list(|r| {
                    let rest = r.bool()?;
                    Ok((rest, r.binding()?))
                })?;
                Ok(Binding::Array { span, elements })
            }
            3 => {
                let entries = self.list(|r| {
                    let key = r.str()?;
                    Ok((key, r.binding()?))
                })?;
                Ok(Binding::Map { span, entries })
            }
            other => Err(format!("unknown binding tag {other}")),
        }
    }

    fn pattern(&mut self) -> Result<Pattern> {
        let span = self.span()?;
        match self.u8()? {
            1 => Ok(Pattern::Wildcard { span }),
            2 => Ok(Pattern::Binding { span, name: self.str()? }),
            3 => Ok(Pattern::Literal { span, value: self.expr()? }),
            other => Err(format!("unknown pattern tag {other}")),
        }
    }

    fn expr(&mut self) -> Result<Expr> {
        let span = self.span()?;
        Ok(match self.u8()? {
            1 => Expr::Number { span, value: self.f64()? },
            2 => Expr::Str {
                span,
                parts: self.list(|r| {
                    Ok(if r.u8()? == 0 {
                        StringPart::Text(r.str()?)
                    } else {
                        StringPart::Expr(r.expr()?)
                    })
                })?,
            },
            3 => Expr::Bool { span, value: self.bool()? },
            4 => Expr::Nil { span },
            5 => Expr::Ident { span, name: self.str()? },
            6 => Expr::Array { span, elements: self.list(|r| r.expr())? },
            7 => Expr::Map {
                span,
                entries: self.list(|r| {
                    let key = r.expr()?;
                    Ok((key, r.expr()?))
                })?,
            },
            8 => {
                let name = self.str()?;
                let params = self.params()?;
                Expr::Fn { span, name, params, body: self.block()? }
            }
            9 => {
                let op = match self.u8()? {
                    0 => UnaryOp::Neg,
                    1 => UnaryOp::Not,
                    other => return Err(format!("unknown unary operator {other}")),
                };
                Expr::Unary { span, op, operand: Box::new(self.expr()?) }
            }
            10 => {
                let tag = self.u8()?;
                let op = BinaryOp::from_tag(tag).ok_or_else(|| format!("unknown binary operator {tag}"))?;
                let op_span = self.span()?;
                let left = Box::new(self.expr()?);
                Expr::Binary { span, op, op_span, left, right: Box::new(self.expr()?) }
            }
            11 => {
                let op = match self.u8()? {
                    0 => LogicalOp::And,
                    1 => LogicalOp::Or,
                    2 => LogicalOp::Coalesce,
                    other => return Err(format!("unknown logical operator {other}")),
                };
                let left = Box::new(self.expr()?);
                Expr::Logical { span, op, left, right: Box::new(self.expr()?) }
            }
            12 => {
                let op = match self.u8()? {
                    0 => None,
                    1 => Some(BinaryOp::Add),
                    2 => Some(BinaryOp::Sub),
                    3 => Some(BinaryOp::Mul),
                    4 => Some(BinaryOp::Div),
                    5 => Some(BinaryOp::Rem),
                    other => return Err(format!("unknown assignment operator {other}")),
                };
                let target = Box::new(self.expr()?);
                Expr::Assign { span, op, target, value: Box::new(self.expr()?) }
            }
            13 => {
                let args_span = self.span()?;
                let callee = Box::new(self.expr()?);
                Expr::Call { span, args_span, callee, args: self.list(|r| r.expr())? }
            }
            14 => {
                let property = self.str()?;
                let property_span = self.span()?;
                Expr::Member { span, property, property_span, object: Box::new(self.expr()?) }
            }
            15 => {
                let object = Box::new(self.expr()?);
                Expr::Index { span, object, index: Box::new(self.expr()?) }
            }
            16 => {
                let inclusive = self.bool()?;
                let start = Box::new(self.expr()?);
                Expr::Range { span, inclusive, start, end: Box::new(self.expr()?) }
            }
            17 => {
                let subject = Box::new(self.expr()?);
                let arms = self.list(|r| {
                    let span = r.span()?;
                    let patterns = r.list(|p| p.pattern())?;
                    let guard = if r.u8()? == 0 { None } else { Some(r.expr()?) };
                    Ok(MatchArm { span, patterns, guard, body: r.expr()? })
                })?;
                Expr::Match { span, subject, arms }
            }
            other => return Err(format!("unknown expression tag {other}")),
        })
    }

    fn stmt(&mut self) -> Result<Stmt> {
        let span = self.span()?;
        Ok(match self.u8()? {
            1 => {
                let mutable = self.bool()?;
                let exported = self.bool()?;
                let binding = self.binding()?;
                Stmt::Let { span, mutable, exported, binding, value: self.expr()? }
            }
            2 => {
                let name = self.str()?;
                let exported = self.bool()?;
                let params = self.params()?;
                let body = self.block()?;
                Stmt::Fn(Rc::new(FnDecl { name, exported, params, body, span }))
            }
            3 => Stmt::Expr { span, expr: self.expr()? },
            4 => Stmt::Baa { span, values: self.list(|r| r.expr())? },
            5 => {
                let value = if self.u8()? == 0 { None } else { Some(self.expr()?) };
                Stmt::Return { span, value }
            }
            6 => {
                let condition = self.expr()?;
                let consequent = self.block()?;
                let alternate = match self.u8()? {
                    0 => None,
                    1 => Some(Box::new(Else::Block(self.block()?))),
                    2 => Some(Box::new(Else::If(self.stmt()?))),
                    other => return Err(format!("unknown else tag {other}")),
                };
                Stmt::If { span, condition, consequent, alternate }
            }
            7 => {
                let condition = self.expr()?;
                Stmt::While { span, condition, body: self.block()? }
            }
            8 => {
                let name = self.str()?;
                let value_name = if self.u8()? == 0 { None } else { Some(self.str()?) };
                let iterable = self.expr()?;
                Stmt::For { span, name, value_name, iterable, body: self.block()? }
            }
            9 => Stmt::Break { span },
            10 => Stmt::Continue { span },
            11 => {
                let target = if self.u8()? == 1 {
                    ImportTarget::Module(self.u32()? as usize)
                } else {
                    ImportTarget::Std(self.str()?)
                };
                let alias = self.str()?;
                let source_span = self.span()?;
                let named = self.list(|r| {
                    let name = r.str()?;
                    let alias = r.str()?;
                    Ok(ImportSpec { name, alias, span: r.span()? })
                })?;
                Stmt::Import { span, target, alias, source_span, named }
            }
            12 => Stmt::Throw { span, value: self.expr()? },
            13 => {
                let block = self.block()?;
                let handler = if self.u8()? == 0 {
                    None
                } else {
                    let name = if self.u8()? == 0 { None } else { Some(self.str()?) };
                    Some((name, self.block()?))
                };
                let finalizer = if self.u8()? == 0 { None } else { Some(self.block()?) };
                Stmt::Try { span, block, handler, finalizer }
            }
            14 => {
                let name = self.str()?;
                Stmt::Test { span, name, body: self.block()? }
            }
            other => return Err(format!("unknown statement tag {other}")),
        })
    }
}

/// Decodes an image, or explains why the bytes are not one.
pub fn decode(bytes: &[u8]) -> Result<Image> {
    if bytes.len() < MAGIC.len() + 1 || &bytes[..MAGIC.len()] != MAGIC {
        return Err("this is not a Baa image: the header is missing".to_string());
    }
    let version = bytes[MAGIC.len()];
    if version != VERSION {
        return Err(format!(
            "this image is version {version}, and this runtime reads version {VERSION}. \
             Rebuild the application with a matching `baa` release."
        ));
    }

    let mut reader = Reader { bytes, at: MAGIC.len() + 1, strings: Vec::new() };

    let count = reader.count()?;
    let mut strings = Vec::with_capacity(count);
    for _ in 0..count {
        let length = reader.u32()? as usize;
        let end = reader.at + length;
        let slice = reader
            .bytes
            .get(reader.at..end)
            .ok_or("image ends in the middle of a string")?;
        reader.at = end;
        strings.push(Rc::from(
            std::str::from_utf8(slice).map_err(|_| "a string in the image is not valid UTF-8")?,
        ));
    }
    reader.strings = strings;

    let app = reader.list(|r| {
        let key = r.str()?;
        Ok((key, r.str()?))
    })?;
    let entry = reader.u32()? as usize;
    let modules = reader.list(|r| {
        let name = r.str()?;
        let path = r.str()?;
        let source = r.str()?;
        Ok(Module { name, path, source, body: r.block()? })
    })?;

    if entry >= modules.len() {
        return Err(format!(
            "the image names module {entry} as its entry point, but holds {} modules",
            modules.len()
        ));
    }
    for module in &modules {
        check_import_targets(&module.body, modules.len())?;
    }

    Ok(Image { modules, entry, app })
}

/// Every module index in the image points at a module in the image.
///
/// Checked once, here, so the interpreter can index the module list without a
/// bounds test on every import and without a panic if one is wrong.
fn check_import_targets(body: &[Stmt], count: usize) -> Result<()> {
    for statement in body {
        match statement {
            Stmt::Import { target: ImportTarget::Module(index), .. } if *index >= count => {
                return Err(format!("an import points at module {index}, which the image does not hold"));
            }
            Stmt::If { consequent, alternate, .. } => {
                check_import_targets(consequent, count)?;
                match alternate.as_deref() {
                    Some(Else::Block(block)) => check_import_targets(block, count)?,
                    Some(Else::If(statement)) => check_import_targets(std::slice::from_ref(statement), count)?,
                    None => {}
                }
            }
            Stmt::While { body, .. } | Stmt::For { body, .. } | Stmt::Test { body, .. } => {
                check_import_targets(body, count)?;
            }
            Stmt::Try { block, handler, finalizer, .. } => {
                check_import_targets(block, count)?;
                if let Some((_, handler)) = handler {
                    check_import_targets(handler, count)?;
                }
                if let Some(finalizer) = finalizer {
                    check_import_targets(finalizer, count)?;
                }
            }
            _ => {}
        }
    }
    Ok(())
}

/// An image appended to this executable, if there is one.
///
/// A built application is the runtime followed by its image and a footer, so
/// one file is both the program and the thing that runs it. Reading our own
/// path rather than looking beside it means an application does not change
/// behaviour based on what is in its directory.
pub fn embedded(exe: &[u8]) -> Option<&[u8]> {
    const FOOTER: &[u8] = b"BAAFLEECE";
    if exe.len() < FOOTER.len() + 8 {
        return None;
    }
    let tail = &exe[exe.len() - FOOTER.len()..];
    if tail != FOOTER {
        return None;
    }
    let length_at = exe.len() - FOOTER.len() - 8;
    let mut buffer = [0u8; 8];
    buffer.copy_from_slice(&exe[length_at..length_at + 8]);
    let length = u64::from_le_bytes(buffer) as usize;
    if length > length_at {
        return None;
    }
    Some(&exe[length_at - length..length_at])
}

File: src/native/image.ts
/**
 * The `.fleece` image: a Baa program, parsed and checked, in a form a native
 * runtime can load without a parser.
 *
 * Why a serialised tree rather than source or bytecode.
 *
 * Source would mean a second lexer and parser living in the native runtime,
 * and two frontends drift. Every diagnostic, every precedence rule and every
 * newline decision would have to be maintained twice, and the second copy
 * would be the one nobody runs `baa check` through. An image keeps exactly one
 * frontend: the reference implementation in `src/`, which the whole test suite
 * already covers.
 *
 * Bytecode would mean a compiler as well, and a compiler is the part most
 * likely to disagree with the tree-walker about a corner of the semantics.
 * That is worth doing when a measurement asks for it (see ARCHITECTURE.md); it
 * is not the thing to build before anything runs at all.
 *
 * The format is versioned and the runtime refuses an image it was not built
 * for, because a silently misread tree is a wrong answer rather than an error.
 */

import type {
  Binding,
  Block,
  Expression,
  MatchArm,
  Param,
  Pattern,
  Statement,
} from "../ast/ast.ts";
import type { Span } from "../diagnostics/source.ts";

export const MAGIC = "FLEECE\n";
export const IMAGE_VERSION = 1;

/** A module in the bundle, already parsed and checked. */
export type ImageModule = {
  /** Import name, for diagnostics: the file's stem. */
  readonly name: string;
  /** Path as written into the image, relative to the project root. */
  readonly path: string;
  /** Original text, so a runtime error can underline the line it happened on. */
  readonly source: string;
  readonly body: readonly Statement[];
};

export type Image = {
  readonly modules: readonly ImageModule[];
  /** Index into `modules` of the program that runs first. */
  readonly entry: number;
  /** Application metadata, from `[app]` in `baa.toml`. */
  readonly app: Readonly<Record<string, string>>;
  /** Resolves a relative import to an index in `modules`. */
  readonly resolveImport: (fromModule: number, source: string) => number;
  /** True when this import was bundled: a relative path, or a dependency. */
  readonly hasImport: (fromModule: number, source: string) => boolean;
};

// Statement tags. The numbers are part of the format: append, never renumber.
const S_LET = 1;
const S_FN = 2;
const S_EXPR = 3;
const S_BAA = 4;
const S_RETURN = 5;
const S_IF = 6;
const S_WHILE = 7;
const S_FOR = 8;
const S_BREAK = 9;
const S_CONTINUE = 10;
const S_IMPORT = 11;
const S_THROW = 12;
const S_TRY = 13;
const S_TEST = 14;

const E_NUMBER = 1;
const E_STRING = 2;
const E_BOOL = 3;
const E_NIL = 4;
const E_IDENT = 5;
const E_ARRAY = 6;
const E_MAP = 7;
const E_FN = 8;
const E_UNARY = 9;
const E_BINARY = 10;
const E_LOGICAL = 11;
const E_ASSIGN = 12;
const E_CALL = 13;
const E_MEMBER = 14;
const E_INDEX = 15;
const E_RANGE = 16;
const E_MATCH = 17;

const B_NAME = 1;
const B_ARRAY = 2;
const B_MAP = 3;

const P_WILDCARD = 1;
const P_BINDING = 2;
const P_LITERAL = 3;

/**
 * Operators travel as small integers rather than strings, so the runtime
 * matches on a number instead of comparing text on every arithmetic operation.
 */
const BINARY_OPS = ["+", "-", "*", "/", "%", "**", "==", "!=", "<", "<=", ">", ">=", "in"] as const;
const LOGICAL_OPS = ["&&", "||", "??"] as const;
const ASSIGN_OPS = ["=", "+=", "-=", "*=", "/=", "%="] as const;
const UNARY_OPS = ["-", "!"] as const;

class Writer {
  readonly bytes: number[] = [];
  readonly #strings = new Map<string, number>();
  readonly stringList: string[] = [];

  u8(value: number): void {
    this.bytes.push(value & 0xff);
  }

  u32(value: number): void {
    this.bytes.push(
      value & 0xff,
      (value >>> 8) & 0xff,
      (value >>> 16) & 0xff,
      (value >>> 24) & 0xff,
    );
  }

  f64(value: number): void {
    const buffer = new ArrayBuffer(8);
    new DataView(buffer).setFloat64(0, value, true);
    for (const byte of new Uint8Array(buffer)) this.bytes.push(byte);
  }

  bool(value: boolean): void {
    this.u8(value ? 1 : 0);
  }

  /** Strings are interned: a program repeats identifiers more than anything else. */
  str(text: string): void {
    let index = this.#strings.get(text);
    if (index === undefined) {
      index = this.stringList.length;
      this.#strings.set(text, index);
      this.stringList.push(text);
    }
    this.u32(index);
  }

  span(span: Span): void {
    this.u32(span.start);
    this.u32(span.end);
  }

  list<T>(items: readonly T[], write: (item: T) => void): void {
    this.u32(items.length);
    for (const item of items) write(item);
  }
}

export function encodeImage(image: Image): Uint8Array {
  const w = new Writer();
  let currentModule = 0;

  const writeBlock = (block: Block): void => {
    w.list(block.body, writeStatement);
  };

  const writeParams = (params: readonly Param[]): void => {
    w.list(params, (param) => {
      w.str(param.name);
      w.bool(param.rest);
      if (param.defaultValue === null) {
        w.u8(0);
      } else {
        w.u8(1);
        writeExpression(param.defaultValue);
      }
    });
  };

  const writeBinding = (binding: Binding): void => {
    w.span(binding.span);
    switch (binding.kind) {
      case "NameBinding":
        w.u8(B_NAME);
        w.str(binding.name);
        return;
      case "ArrayBinding":
        w.u8(B_ARRAY);
        w.list(binding.elements, (element) => {
          w.bool(element.rest);
          writeBinding(element.binding);
        });
        return;
      case "MapBinding":
        w.u8(B_MAP);
        w.list(binding.entries, (entry) => {
          w.str(entry.key);
          writeBinding(entry.binding);
        });
        return;
    }
  };

  const writePattern = (pattern: Pattern): void => {
    w.span(pattern.span);
    switch (pattern.kind) {
      case "WildcardPattern":
        w.u8(P_WILDCARD);
        return;
      case "BindingPattern":
        w.u8(P_BINDING);
        w.str(pattern.name);
        return;
      case "LiteralPattern":
        w.u8(P_LITERAL);
        writeExpression(pattern.value);
        return;
    }
  };

  const writeArm = (arm: MatchArm): void => {
    w.span(arm.span);
    w.list(arm.patterns, writePattern);
    if (arm.guard === null) {
      w.u8(0);
    } else {
      w.u8(1);
      writeExpression(arm.guard);
    }
    writeExpression(arm.body);
  };

  function writeExpression(expression: Expression): void {
    w.span(expression.span);
    switch (expression.kind) {
      case "IntLiteral":
      case "FloatLiteral":
        w.u8(E_NUMBER);
        w.f64(expression.value);
        return;
      case "StringLiteral":
        w.u8(E_STRING);
        w.list(expression.segments, (segment) => {
          if (segment.kind === "text") {
            w.u8(0);
            w.str(segment.value);
          } else {
            w.u8(1);
            writeExpression(segment.expression);
          }
        });
        return;
      case "BoolLiteral":
        w.u8(E_BOOL);
        w.bool(expression.value);
        return;
      case "NilLiteral":
        w.u8(E_NIL);
        return;
      case "Identifier":
        w.u8(E_IDENT);
        w.str(expression.name);
        return;
      case "ArrayLiteral":
        w.u8(E_ARRAY);
        w.list(expression.elements, writeExpression);
        return;
      case "MapLiteral":
        w.u8(E_MAP);
        w.list(expression.entries, (entry) => {
          writeExpression(entry.key);
          writeExpression(entry.value);
        });
        return;
      case "FunctionExpression":
        w.u8(E_FN);
        w.str(expression.name ?? "anonymous");
        writeParams(expression.params);
        writeBlock(expression.body);
        return;
      case "UnaryExpression":
        w.u8(E_UNARY);
        w.u8(UNARY_OPS.indexOf(expression.operator));
        writeExpression(expression.operand);
        return;
      case "BinaryExpression":
        w.u8(E_BINARY);
        w.u8(BINARY_OPS.indexOf(expression.operator));
        w.span(expression.operatorSpan);
        writeExpression(expression.left);
        writeExpression(expression.right);
        return;
      case "LogicalExpression":
        w.u8(E_LOGICAL);
        w.u8(LOGICAL_OPS.indexOf(expression.operator));
        writeExpression(expression.left);
        writeExpression(expression.right);
        return;
      case "AssignmentExpression":
        w.u8(E_ASSIGN);
        w.u8(ASSIGN_OPS.indexOf(expression.operator));
        writeExpression(expression.target);
        writeExpression(expression.value);
        return;
      case "CallExpression":
        w.u8(E_CALL);
        w.span(expression.argsSpan);
        writeExpression(expression.callee);
        w.list(expression.args, writeExpression);
        return;
      case "MemberExpression":
        w.u8(E_MEMBER);
        w.str(expression.property);
        w.span(expression.propertySpan);
        writeExpression(expression.object);
        return;
      case "IndexExpression":
        w.u8(E_INDEX);
        writeExpression(expression.object);
        writeExpression(expression.index);
        return;
      case "RangeExpression":
        w.u8(E_RANGE);
        w.bool(expression.inclusive);
        writeExpression(expression.start);
        writeExpression(expression.end);
        return;
      case "MatchExpression":
        w.u8(E_MATCH);
        writeExpression(expression.subject);
        w.list(expression.arms, writeArm);
        return;
    }
  }

  function writeStatement(statement: Statement): void {
    w.span(statement.span);
    switch (statement.kind) {
      case "LetStatement":
        w.u8(S_LET);
        w.bool(statement.mutable);
        w.bool(statement.exported);
        writeBinding(statement.binding);
        writeExpression(statement.value);
        return;
      case "FunctionDeclaration":
        w.u8(S_FN);
        w.str(statement.name);
        w.bool(statement.exported);
        writeParams(statement.params);
        writeBlock(statement.body);
        return;
      case "ExpressionStatement":
        w.u8(S_EXPR);
        writeExpression(statement.expression);
        return;
      case "BaaStatement":
        w.u8(S_BAA);
        w.list(statement.values, writeExpression);
        return;
      case "ReturnStatement":
        w.u8(S_RETURN);
        if (statement.value === null) {
          w.u8(0);
        } else {
          w.u8(1);
          writeExpression(statement.value);
        }
        return;
      case "IfStatement":
        w.u8(S_IF);
        writeExpression(statement.condition);
        writeBlock(statement.consequent);
        if (statement.alternate === null) {
          w.u8(0);
        } else if (statement.alternate.kind === "Block") {
          w.u8(1);
          writeBlock(statement.alternate);
        } else {
          // `else if` stays a statement rather than being flattened into a
          // block, so the runtime reports the span the source actually has.
          w.u8(2);
          writeStatement(statement.alternate);
        }
        return;
      case "WhileStatement":
        w.u8(S_WHILE);
        writeExpression(statement.condition);
        writeBlock(statement.body);
        return;
      case "ForStatement":
        w.u8(S_FOR);
        w.str(statement.name);
        if (statement.valueName === null) {
          w.u8(0);
        } else {
          w.u8(1);
          w.str(statement.valueName);
        }
        writeExpression(statement.iterable);
        writeBlock(statement.body);
        return;
      case "BreakStatement":
        w.u8(S_BREAK);
        return;
      case "ContinueStatement":
        w.u8(S_CONTINUE);
        return;
      case "ImportDeclaration": {
        w.u8(S_IMPORT);
        // Resolved here, not at runtime: the native runtime never consults the
        // filesystem to find a module, so a shipped application cannot be
        // redirected by a file dropped next to it.
        // A `[wool]` dependency is bundled like any other module, so it is
        // written as an index too: `import my_lib` and `import "./lib.baa"`
        // differ only in how they were spelled.
        const bundled = statement.relative ? true : image.hasImport(currentModule, statement.source);
        if (bundled) {
          w.u8(1);
          w.u32(image.resolveImport(currentModule, statement.source));
        } else {
          w.u8(0);
          w.str(statement.source);
        }
        w.str(statement.alias);
        w.span(statement.sourceSpan);
        w.list(statement.named, (specifier) => {
          w.str(specifier.name);
          w.str(specifier.alias);
          w.span(specifier.span);
        });
        return;
      }
      case "ThrowStatement":
        w.u8(S_THROW);
        writeExpression(statement.value);
        return;
      case "TryStatement":
        w.u8(S_TRY);
        writeBlock(statement.block);
        if (statement.handler === null) {
          w.u8(0);
        } else {
          w.u8(1);
          if (statement.catchName === null) {
            w.u8(0);
          } else {
            w.u8(1);
            w.str(statement.catchName);
          }
          writeBlock(statement.handler);
        }
        if (statement.finalizer === null) {
          w.u8(0);
        } else {
          w.u8(1);
          writeBlock(statement.finalizer);
        }
        return;
      case "TestDeclaration":
        w.u8(S_TEST);
        w.str(statement.name);
        writeBlock(statement.body);
        return;
    }
  }

  const appKeys = Object.keys(image.app).sort();
  w.list(appKeys, (key) => {
    w.str(key);
    w.str(image.app[key]!);
  });

  w.u32(image.entry);
  w.list(image.modules, (module, ...rest) => {
    void rest;
    currentModule = image.modules.indexOf(module);
    w.str(module.name);
    w.str(module.path);
    w.str(module.source);
    w.list(module.body, writeStatement);
  });

  // The string table is only complete once the body has been written, so it is
  // prepended here rather than reserved and patched.
  const head: number[] = [...new TextEncoder().encode(MAGIC), IMAGE_VERSION];
  const encoder = new TextEncoder();
  const table: number[] = [];
  const count = w.stringList.length;
  table.push(count & 0xff, (count >>> 8) & 0xff, (count >>> 16) & 0xff, (count >>> 24) & 0xff);
  for (const text of w.stringList) {
    const encoded = encoder.encode(text);
    const length = encoded.length;
    table.push(length & 0xff, (length >>> 8) & 0xff, (length >>> 16) & 0xff, (length >>> 24) & 0xff);
    for (const byte of encoded) table.push(byte);
  }
  return new Uint8Array([...head, ...table, ...w.bytes]);
}
